### PR TITLE
drivers/flash: Enable flash driver for stm32f2

### PR DESCRIPTION
**Pull request by ikmdani**
This commit enables flash driver for nucleo_f207zg platform.
This has been tested with tests/drivers/flash.

### CODE REVIEW

File: drivers/flash/flash_stm32f2x.c [hunk 1]
```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2021 Linaro Limited
```
Comment (erwango) [inline drivers/flash/flash_stm32f2x.c:2]: STMicroelectronics

Comment (erwango) [inline drivers/flash/flash_stm32f2x.c:2]: STMicroelectronics

File: drivers/flash/flash_stm32f2x.c [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+ * Different SoC flash layouts are specified in across various
+ * reference manuals, but the flash layout for a given number of
+ * sectors is consistent across these manuals, with one "gotcha". The
+ * number of sectors is given by the HAL as FLASH_SECTOR_TOTAL.
+ *
+ * The only "gotcha" is that when there are 24 sectors, they are split
+ * across 2 "banks" of 12 sectors each, with another set of small
+ * sectors (16 KB) in the second bank occurring after the large ones
+ * (128 KB) in the first. We could consider supporting this as two
+ * devices to make the layout cleaner, but this will do for now.
```
Comment (erwango) [inline drivers/flash/flash_stm32f2x.c:142]: Update this section. Remove if not needed.

File: drivers/flash/flash_stm32f2x.c [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	return rc;
+}
+
+/*
+ * Different SoC flash layouts are specified in across various
+ * reference manuals, but the flash layout for a given number of
+ * sectors is consistent across these manuals, with one "gotcha". The
+ * number of sectors is given by the HAL as FLASH_SECTOR_TOTAL.
```
Comment (erwango) [inline drivers/flash/flash_stm32f2x.c:136]: From what I can see there is only one flash layout on F2 series. This section would require an update.

**Review by erwango — CHANGES_REQUESTED** [review]
Few minor comments, otherwise LGTM

File: boards/arm/nucleo_f207zg/nucleo_f207zg.dts [hunk 4]
```diff
@@ -130,6 +130,20 @@
...
+&flash0 {
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		/* Set 2KB of storage at the end of 1MB flash */
+		storage_partition: partition@ff800 {
+			label = "storage";
+			reg = <0x000FF800 0x00000800>;
```
Comment (FRASTM) [inline boards/arm/nucleo_f207zg/nucleo_f207zg.dts:142]: usually in lower case : 0x000**ff**800

**Comment by FRASTM** [comment]
running the tests/drivers/flash running on the nucleo_f207zg

**Review by erwango — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]

File: drivers/flash/flash_stm32f2x.c [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+#include <kernel.h>
+#include <device.h>
+#include <string.h>
+#include <drivers/flash.h>
+#include <init.h>
+#include <soc.h>
+
+#include "flash_stm32.h"
+
+#define STM32F2X_SECTOR_MASK		((uint32_t) 0xFFFFFF07)
```
Comment (erwango) [inline drivers/flash/flash_stm32f2x.c:16]: This shouldn't be needed, cf https://github.com/zephyrproject-rtos/zephyr/pull/36144

**Review by erwango — CHANGES_REQUESTED** [review]

**Review by erwango — APPROVED** [review]

**Review by nvlsianpu — APPROVED** [review]

File: boards/arm/nucleo_f207zg/nucleo_f207zg.dts [hunk 6]
```diff
@@ -130,6 +130,20 @@
 		     &eth_txd1_pb13>;
 };
 
+&flash0 {
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		/* Set 2KB of storage at the end of 1MB flash */
```
Comment (ABOSTM) [inline boards/arm/nucleo_f207zg/nucleo_f207zg.dts:139]: The last sector of STM32F2xx is 128KB not 2KB

File: boards/arm/nucleo_f207zg/nucleo_f207zg.dts [hunk 7]
```diff
@@ -130,6 +130,20 @@
...
+		storage_partition: partition@ff800 {
+			label = "storage";
+			reg = <0x000ff800 0x00000800>;
```
Comment (ABOSTM) [inline boards/arm/nucleo_f207zg/nucleo_f207zg.dts:142]: The last sector of STM32F2xx is 128KB not 2KB

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: boards/arm/nucleo_f207zg/nucleo_f207zg.dts (hunk 6 above)
Comment (erwango) [reply boards/arm/nucleo_f207zg/nucleo_f207zg.dts:139]: Good catch, thanks @ABOSTM

**Comment by FRASTM** [comment]
west build -p auto -b nucleo_f207zg tests/drivers/flash

```
*** Booting Zephyr OS build zephyr-v2.6.0-119-g87c45282ea33 ***
 Running test suite flash_driver_test
 ===================================================================
 START - test_setup
 PASS - test_setup in 0.1 seconds
 ===================================================================
 START - test_read_unaligned_address
 PASS - test_read_unaligned_address in 0.8 seconds
 ===================================================================
 Test suite flash_driver_test succeeded
 ===================================================================
 PROJECT EXECUTION SUCCESSFUL
```

**Comment by ikmdani** [comment]
$ ./scripts/twister --device-testing --device-serial /dev/ttyACM0 -p nucleo_f207zg -j 2 -T tests/drivers/flash
```
ZEPHYR_BASE unset, using "/home/sarthsmart/zephyrproject/zephyr"
Renaming output directory to /home/sarthsmart/zephyrproject/zephyr/twister-out.17
INFO    - Zephyr version: zephyr-v2.6.0-282-gda906cca2403
INFO    - JOBS: 2
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testcase list...
INFO    - 4 test scenarios (4 configurations) selected, 3 configurations discarded due to filters.

 

Device testing on:

 

| Platform      | ID   | Serial device   |
|---------------|------|-----------------|
| nucleo_f207zg |      | /dev/ttyACM0    |

 

INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:    1/   1  100%  skipped:    4, failed:    0
INFO    - 0 of 0 test configurations passed (0.00%), 0 failed, 4 skipped with 0 warnings in 7.93 seconds
INFO    - In total 0 test cases were executed, 8 skipped on 1 out of total 370 platforms (0.27%)
INFO    - 0 test configurations executed on platforms, 0 test configurations were only built.

 

Hardware distribution summary:

 

| Board         | ID   |   Counter |
|---------------|------|-----------|
| nucleo_f207zg |      |         0 |
INFO    - Saving reports...
INFO    - Writing xunit report /home/sarthsmart/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/sarthsmart/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

**Comment by ABOSTM** [comment]
@ikmdani , as I just pushed https://github.com/zephyrproject-rtos/zephyr/pull/36277
where I enabled ART flash cache accelerator,
can you add cache coherency management for this driver ?
It is similar to what I have done for stm32f4 in https://github.com/zephyrproject-rtos/zephyr/pull/36277
also very similar to merged PR (for stm32g4) https://github.com/zephyrproject-rtos/zephyr/pull/32218

**Comment by ikmdani** [comment]
> can you add cache coherency management for this driver ?

Done, raised updated PR, kindly review.

**Comment by ABOSTM** [comment]
> > can you add cache coherency management for this driver ?
> 
> Done, raised updated PR, kindly review.

Thanks

**Review by ABOSTM — APPROVED** [review]
LGTM

**Pull request merged**